### PR TITLE
#4597: tee-stdio capture method

### PR DESCRIPTION
**Pull request by csm10495**
- Thanks for submitting a PR, your contribution is really appreciated!

- Here is a quick checklist that should be present in PRs.

- [x] Target the `features` branch for new features, improvements, and removals/deprecations.
- [x] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.

- Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/master/changelog/README.rst) for details.

Add tee-stdio capture mechanism to allow both live printing and capturing of test output.

- [x] Add yourself to `AUTHORS` in alphabetical order.

### CODE REVIEW

File: doc/en/capture.rst [hunk 1]
```diff
@@ -33,15 +33,22 @@ There are two ways in which ``pytest`` can perform capturing:
   and ``sys.stderr`` will be captured.  No capturing of writes to
   filedescriptors is performed.
 
+* tee-stdio capturing: Python writes to ``sys.stdout`` and ``sys.stderr``
+  will be captured, however the writes will also be passed-through to
+  the actual ``sys.stdout`` and ``sys.stderr``. This allows output to be
+  'live printed' and captured for plugin use, such as junitxml.
```
Comment (nicoddemus) [inline doc/en/capture.rst]:
```suggestion
  'live printed' and captured for plugin use, such as junitxml (new in pytest 5.4).
```

File: doc/en/capture.rst [hunk 2]
```diff
@@ -24,7 +24,7 @@ a test.
 Setting capturing methods or disabling capturing
 -------------------------------------------------
 
-There are two ways in which ``pytest`` can perform capturing:
+There are three ways in which ``pytest`` can perform capturing:
 
 * file descriptor (FD) level capturing (default): All writes going to the
```
Comment (nicoddemus) [inline doc/en/capture.rst]:
```suggestion
* ``fd`` (file descriptor) level capturing (default): All writes going to the
```

File: doc/en/capture.rst [hunk 3]
```diff
@@ -33,15 +33,22 @@ There are two ways in which ``pytest`` can perform capturing:
   and ``sys.stderr`` will be captured.  No capturing of writes to
   filedescriptors is performed.
 
+* tee-stdio capturing: Python writes to ``sys.stdout`` and ``sys.stderr``
```
Comment (nicoddemus) [inline doc/en/capture.rst]:
```suggestion
* ``tee-stdio`` capturing: Python writes to ``sys.stdout`` and ``sys.stderr``
```

File: testing/test_capture.py [hunk 4]
```diff
@@ -1112,6 +1136,22 @@ def test_stdin_nulled_by_default(self):
             pytest.raises(IOError, sys.stdin.read)
 
 
+class TestTeeStdCapture(TestStdCapture):
```
Comment (nicoddemus) [inline testing/test_capture.py]: Good tests!

I suggest we add a test to `acceptance_tests.py` which uses `--capture=tee-stdio` and `--junitxml` and checks that output is shown in the terminal and in the generated XML file, just to ensure we don't regress the original intent of the feature.

Comment (csm10495) [reply testing/test_capture.py]: Great idea! Added one with a045e3f52dcca.

**Comment by csm10495** [comment]
Added the requested test and got all checks passing :)

**Comment by blueyed** [comment]
Yeah, this should definitely be `tee-sys` (if it stays this way).

But I think it makes more sense as a separate option that gets handled by `MultiCapture` then, i.e. the wrapper of `FDCapture`/`SysCapture`.  Then it would work for both methods.

(In #4597 it was mentioned to be more tricky for fds, but it is done here by just writing to the orig handler anyway (which likely also works for the fd method then).)

**Comment by csm10495** [comment]
Alrighty swapped name to tee-sys

**Comment by RonnyPfannschmidt** [comment]
@blueyed fd-tee has to have a completely different solution, as it has to work for non python and subprocesses

**Comment by blueyed** [comment]
My thought/idea was that MultiCapture would just call through to `write_orig` (or whatever it is called), but have not really checked the code etc.

**Comment by RonnyPfannschmidt** [comment]
@blueyed that only works for python level file objects (its not enough for a fd)

**Comment by nicoddemus** [comment]
Thanks again @csm10495!

I've added a link to the docs in the changelog entry and rebased/squashed the commits. 👍 

We should merge once CI passes.

**Comment by ssbarnea** [comment]
Thanks! This is a really useful feature.

**Pull request merged**